### PR TITLE
Convert NewItems plugin to service helper

### PR DESCRIPTION
**Pull request by sambhavp96**

### CODE REVIEW

**Review by demiankatz — CHANGES_REQUESTED** [review]
@sambhavp96, as you may have already noticed, there appears to be an existing unit test for the NewItems helper -- hopefully you can rename and slightly adjust this to make things pass. I'll give this a closer review once that is done; let me know if I can assist with anything else in the meantime.

File: module/VuFind/src/VuFind/Search/NewItemsHelper.php [hunk 1]
```diff
@@ -21,64 +21,78 @@
...
 
     /**
      * Constructor
      *
-     * @param Config $config Configuration
+     * @param Config     $config  Configuration
+     * @param Connection $catalog ILS connection
      */
-    public function __construct(Config $config)
+    public function __construct(Config $config, Connection $catalog)
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Search/NewItemsHelper.php]: Might as well use property promotion here to remove some lines of code:

```suggestion
    public function __construct(protected Config $config, protected Connection $catalog)
```

File: module/VuFind/src/VuFind/Search/NewItemsHelper.php [hunk 2]
```diff
@@ -21,64 +21,78 @@
...
     {
         $this->config = $config;
+        $this->catalog = $catalog;
     }
 
     /**
      * Figure out which bib IDs to load from the ILS.
      *
-     * @param \VuFind\ILS\Connection     $catalog ILS connection
+     * @param Connection                 $catalog ILS connection (for backward compatibility)
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Search/NewItemsHelper.php]: I think it's safe to drop this parameter and use `$this->connection` instead. Not counting the test, it's only called in one place in the SearchController, and that instance would never use a different connection object than the one that's already been dependency-injected.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Controller/Plugin/NewItemsTest.php [hunk 3]
```diff
@@ -5,32 +5,24 @@
...
  *
- * This program is distributed in the hope that it will be useful,
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- * GNU General Public License for more details.
- *
- * You should have received a copy of the GNU General Public License
- * along with this program; if not, see
- * <https://www.gnu.org/licenses/>.
- *
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Controller/Plugin/NewItemsTest.php:22]: Did you intentionally remove these lines? I don't think I see a reason why they should be deleted.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Search/NewItemsHelperTest.php [hunk 4]
```diff
@@ -216,15 +214,14 @@ public function testGetSolrFilter()
      */
     protected function getMockCatalog(): \VuFind\ILS\Connection
     {
-        $catalog = $this->createMock(\VuFind\ILS\Connection::class);
+        $catalog = $this->createMock(Connection::class);
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Search/NewItemsHelperTest.php:218]: It appears to be an existing problem, but I see a couple of lines below this, line 220 isn't indented correctly. I think the easiest solution would be to combine lines 219-220 into a single line, and then the subsequent indentation will be correct.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Search/NewItemsTest.php [hunk 5]
```diff
@@ -52,7 +44,8 @@ public function testGetBibIDsFromCatalog()
     {
         $flash = $this->createMock(\Laminas\Mvc\Plugin\FlashMessenger\FlashMessenger::class);
         $config = new Config(['result_pages' => 10]);
-        $newItems = new NewItems($config);
+        $catalog = $this->createMock(Connection::class);
+        $newItems = new NewItemsHelper($config, $catalog);
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Search/NewItemsTest.php]: Maybe it's worth creating a support method like:

```
protected function getMockHelper(array $config = [], ?Connection $connection = null)
{
    return new NewItemsHelper(new Config($config), $connection ?? $this->createMock(Connection::class));
}
```

That will reduce some redundancy throughout the file, and give us a convenient central place to make changes if we change the constructor of the helper again in the future. (And that may happen sooner rather than later; see my last suggestion below...)

File: module/VuFind/config/module.config.php [hunk 6]
```diff
@@ -550,6 +548,7 @@
             'VuFind\Search\Factory\UrlQueryHelperFactory' => 'Laminas\ServiceManager\Factory\InvokableFactory',
             'VuFind\Search\History' => 'VuFind\Search\HistoryFactory',
             'VuFind\Search\Memory' => 'VuFind\Search\MemoryFactory',
+            'VuFind\Search\NewItemsHelper' => 'VuFind\Controller\Plugin\NewItemsFactory',
```
Comment (demiankatz) [inline module/VuFind/config/module.config.php]: The factory should be moved into the same namespace as the helper.

File: module/VuFind/src/VuFind/Search/NewItemsHelper.php [hunk 7]
```diff
@@ -21,64 +21,78 @@
...
-class NewItems extends AbstractPlugin
+class NewItemsHelper
 {
     /**
      * Configuration
      *
      * @var Config
      */
-    protected $config;
+    protected Config $config;
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Search/NewItemsHelper.php]: We're working on phasing out Config objects and just using arrays instead. Maybe refactoring this class is an opportunity to remove Config objects from one more place. It shouldn't be too difficult to use `getConfigArray` instead of `getConfigObject` in the factory for the helper, and then switch from object syntax to array syntax in all of the references to $this->config.

If you think that's too much change all at once, we could handle it in a follow-up PR after this one is merged... or you can do it now if you prefer. I'll leave it up to you!

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @sambhavp96 -- see below for a few more suggestions!

File: module/VuFind/config/module.config.php [hunk 8]
```diff
@@ -550,6 +548,7 @@
             'VuFind\Search\Factory\UrlQueryHelperFactory' => 'Laminas\ServiceManager\Factory\InvokableFactory',
             'VuFind\Search\History' => 'VuFind\Search\HistoryFactory',
             'VuFind\Search\Memory' => 'VuFind\Search\MemoryFactory',
+            'VuFind\Search\NewItemsHelper' => 'VuFind\Search\NewItemsFactory',
```
Comment (demiankatz) [inline module/VuFind/config/module.config.php]: The factory class should be renamed to NewItemsHelperFactory for consistency (especially since we may rely more heavily on convention-based auto-loading for factories in the future).

File: module/VuFind/src/VuFind/Search/NewItemsHelper.php [hunk 9]
```diff
@@ -21,70 +21,65 @@
...
-
     /**
      * Constructor
      *
-     * @param Config $config Configuration
+     * @param array      $config  Configuration
+     * @param Connection $catalog ILS connection
      */
-    public function __construct(Config $config)
+    public function __construct(protected array $config, protected Connection $catalog)
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Search/NewItemsHelper.php:56]: It also looks like the factory still needs to be modified use getConfigArray instead of getConfigObject and to inject the ILS Connection object.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Search/NewItemsTest.php [hunk 10]
```diff
@@ -43,6 +43,19 @@
...
 {
+    /**
+     * Get a NewItemsHelper instance for testing.
+     *
+     * @param array           $config     Configuration array
+     * @param Connection|null $connection ILS connection
+     *
+     * @return NewItemsHelper
+     */
+    protected function getMockHelper(array $config = [], ?Connection $connection = null): NewItemsHelper
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Search/NewItemsTest.php]: I realize that I suggested the method name `getMockHelper`, but that was silly of me because this is not a Mock. Do you mind search-and-replacing this to `getNewItemsHelper` instead? :-)

File: module/VuFind/src/VuFind/Search/NewItemsHelper.php [hunk 11]
```diff
@@ -144,16 +138,16 @@ public function getFundList()
      *
      * @return array
      */
-    public function getHiddenFilters()
+    public function getHiddenFilters(): array
     {
-        if (!isset($this->config->filter)) {
+        if (! isset($this->config['filter'])) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Search/NewItemsHelper.php]: Minor style suggestion: we don't put spaces after `!`:

```suggestion
        if (!isset($this->config['filter'])) {
```

Comment (demiankatz) [reply module/VuFind/src/VuFind/Search/NewItemsHelper.php]: Though alternatively, I have a feeling that this entire function could be reduced to:

```
return (array)($this->config['filter'] ?? []);
```

Now that we're not using the objects, a bit of typecasting and null coalescing should cover all the same edge cases -- though feel free to leave it as-is if you don't confidently agree with me and/or if there's not already adequate test coverage of the possible edge cases (string value, array value, no value defined).

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for all the progress, @sambhavp96, this looks increasingly good! There are a couple of significant issues with the factory, though. Is the integration test suite passing for you? If so, we may have some gaps that need addressing there, since I would expect the code to be seriously broken by this -- but if you simply haven't had a chance to run the Mink tests yet, I understand! :-)

**Review by demiankatz — APPROVED** [review]
Thanks, @sambhavp96! During final testing of this, I discovered there was a gap in our automated test coverage of new items, so I opened #4947 to address it. This new test revealed some problems related to types, so I made some minor tweaks here to fix them. I also opened #4949 to correct some problematic annotations in existing code.

I am running the full test suite here one more time. Once it passes, I will merge this PR.
